drivers/virtio/virtio-input-event-codes.h

```diff
@@ -0,0 +1,926 @@
+/****************************************************************************
+ * drivers/virtio/virtio-input-event-codes.h
+ *
+ * Input event codes
+ *
+ *    *** IMPORTANT ***
+ * This file is not only included from C-code but also from devicetree source
+ * files. As such this file MUST only contain comments and defines.
+ *
+ * Copyright (c) 1999-2002 Vojtech Pavlik
+ * Copyright (c) 2015 Hans de Goede <hdegoede@redhat.com>
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation.
```

Comment:
I think this virtio code used here also is coming from Linux!
As I commented here: https://github.com/apache/nuttx/pull/10157 Please don't to that.

Please @patacongo @jerpelea @btashton help us to review it.

Reply:
virtio-input is modeled as Linux input driver, and explicitly reference this header file in the spec:
https://docs.oasis-open.org/virtio/virtio/v1.1/cs01/virtio-v1.1-cs01.html#x1-3390008
and the code comes from include/uapi folder:
https://github.com/torvalds/linux/blob/master/include/uapi/linux/input-event-codes.h
All header files in this folder are used and referenced by many userspace library and application which has a special license:
https://spdx.org/licenses/Linux-syscall-note.html.
All other code is written from scratch. My question is that where is it safe to directly use the interface definition in header files.

Reply:
> My question is that where is it safe to directly use the interface definition in header files.

Caution:  I am not an attorney

My understanding is that is it is okay to use the interface definitions that are normally provided by a GPL linux header file:

https://www.kernel.org/doc/html/v4.18/process/license-rules.html

> The User-space API (UAPI) header files, which describe the interface of user-space programs to the kernel are a special case. According to the note in the kernel COPYING file, the syscall interface is a clear boundary, which does not extend the GPL requirements to any software which uses it to communicate with the kernel. Because the UAPI headers must be includable into any source files which create an executable running on the Linux kernel, the exception must be documented by a special license expression.

The exception is stated in the linux COPYING file:

> NOTE! This copyright does not cover user programs that use kernel services by normal system calls - this is merely considered normal use of the kernel, and does not fall under the heading of "derived work".

But it is not acceptable, however, to copy linux header files into the NuttX repository; the file has an incompatible GPL copyright and cannot be redistributed as Apache code.  I think that the user interface structures can be re-defined in non-linux, Apache header files but that is less clear.  Certainly projects like Wine, for example, have no problem duplicating proprietary Microsoft interface definitions in Wine header files.

It would be safer to design similar, compatible interface and avoid any hint of copy/paste or re-use or even derivation of any Linux header file.

Reply:
@xiaoxiang781216 I think in this case it is better to copy these headers from Android, instead of copying from Linux. 

I'm worried because yesterday GUIDINGLI opened a PR with all the dynamic power frequency source code from Linux, including header files calling #include <linux/blablabla.h>.

Probably it was a mistake, but it also could be that they are using those codes directly as base, that is not allowed by GPL. You cannot look the GPL code and implement your code using it as reference.